copy/multiple_test.go

```diff
@@ -1,36 +1,163 @@
+	require.NoError(t, err)
+	compare = []instanceCopy{{op: instanceCopyCopy,
+		sourceDigest: sourceInstances[1]}}
+	assert.Equal(t, instancesToCopy, compare)
+}
+
+// Test `instanceCopyClone` cases.
+func TestPrepareCopyInstancesforInstanceCopyClone(t *testing.T) {
+	validManifest, err := os.ReadFile(filepath.Join("..", "internal", "manifest", "testdata", "oci1.index.zstd-selection.json"))
+	require.NoError(t, err)
+	list, err := internalManifest.ListFromBlob(validManifest, internalManifest.GuessMIMEType(validManifest))
+	require.NoError(t, err)
+
+	// Prepare option for `instanceCopyClone` case.
+	ensureCompressionVariantsExist := []OptionCompressionVariant{{Algorithm: &compression.Zstd}}
+
+	sourceInstances := []digest.Digest{
+		digest.Digest("sha256:bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb"),
 		digest.Digest("sha256:aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"),
+		digest.Digest("sha256:cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc"),
 	}
-	instancesToCopy = prepareInstanceCopies(sourceInstances, &Options{
-		Instances:          copyOnly,
+
+	// CopySpecificImage must fail with error
+	_, err = prepareInstanceCopies(list, sourceInstances, &Options{EnsureCompressionVariantsExist: ensureCompressionVariantsExist,
+		Instances:          []digest.Digest{sourceInstances[1]},
 		ImageListSelection: CopySpecificImages})
-	assert.Equal(t, instancesToCopy, []instanceCopy{{
-		op:           instanceCopyCopy,
-		sourceDigest: copyOnly[0]}})
+	require.EqualError(t, err, "EnsureCompressionVariant is not implemented for CopySpecificImages")
+
+	// Test copying all images with replication
+	instancesToCopy, err := prepareInstanceCopies(list, sourceInstances, &Options{EnsureCompressionVariantsExist: ensureCompressionVariantsExist})
+	require.NoError(t, err)
+
+	// Following test ensures
+	// * Still copy gzip variants if they exist in the original
+	// * Not create new Zstd variants if they exist in the original.
+
+	// We crated a list of three instances `sourceInstances` and since in oci1.index.zstd-selection.json
+	// amd64 already has a zstd instance i.e sourceInstance[1] so it should not create replication for
+	// `sourceInstance[0]` and `sourceInstance[1]` but should do it for `sourceInstance[2]` for `arm64`
+	// and still copy `sourceInstance[2]`.
+	assert.Equal(t, 4, len(instancesToCopy))
+	expectedResponse := []simplerInstanceCopy{}
+	for _, instance := range sourceInstances {
+		// If its `arm64` and sourceDigest[2] , expect a clone to happen
+		if instance == sourceInstances[2] {
+			expectedResponse = append(expectedResponse, simplerInstanceCopy{op: instanceCopyClone, sourceDigest: instance, cloneCompressionVariant: "zstd", clonePlatform: "arm64-linux-"})
+		}
+		expectedResponse = append(expectedResponse, simplerInstanceCopy{op: instanceCopyCopy,
+			sourceDigest: instance})
+	}
```

Comment:
Absolutely non-blocking: I think using a hard-coded array constant with the expected data would not be longer, and it would actually be a tiny bit easier to follow.

(Going further down that road, all of the new added tests could be table-driven. That would scale better.)

Reply:
Yes, testinfra needs a bit of cleanup. I'll open a follow up PR after this.